### PR TITLE
blk/pmem: Add the devdax support.

### PR DESCRIPTION
**Pull request by optimistyzy**
The purpose is to make the pmem device usage more flexible
than the current solution. And prepare for the potential
offloading by hardware engine later.

Signed-off-by: Ziye Yang <ziye.yang@intel.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/blk/pmem/PMEMDevice.cc
Comment (majianpeng) [inline src/blk/pmem/PMEMDevice.cc:148]: we can still use pmem_map_file for devdax.

File: src/blk/pmem/PMEMDevice.cc
Comment (majianpeng) [inline src/blk/pmem/PMEMDevice.cc:259]: why add those code?

File: src/blk/pmem/PMEMDevice.h
Comment (majianpeng) [inline src/blk/pmem/PMEMDevice.h:44]: Where is this function called?

File: src/blk/pmem/PMEMDevice.cc
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:25]: this file gets included twice.

Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:25]: removed

File: src/blk/pmem/PMEMDevice.cc
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:263]: checking whether it is a devdax file. then use different open flags

File: src/blk/pmem/PMEMDevice.h
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.h:44]: called in Bluefs.cc or BlueStore.cc. For using /dev/dax, it is not a block device. So we cannot allow normal block operations, which is same to use SPDK to drive nvme devices.

**Comment by optimistyzy** [comment]
@tchaikov updated version. With this patch,  we can support using PMEM char device.

File: src/blk/pmem/PMEMDevice.cc
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:68]: please define variable close to where it is used for the first time for better readability.

File: src/blk/pmem/PMEMDevice.cc
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:83]:
```c++
namespace fs = std::filesystem
if (!fs::is_character_file(pmem_file)) {
  return -EINVAL;
}
struct stat file_stat;
if (::fstat(fd, &file_stat)) {
  return -EINVAL;
}
auto char_dir = fmt::format("/sys/dev/char/{}:{}",
                            major(file_stat.st_rdev), 
                            minor(file_stat.st_rdev));
if (auto subsys_path = char_path / "subsystem";
    fs::read_symlink(subsys_path).filename().string() != "dax") {
  return -EINVAL;
}
if (total_size == nullptr) {
  return 0;
}
if (std::ifstream size_file(char_path / "size"); size_file)
  size_file >> *total_size;
  return size_file ? 0 : -EINVAL;
} else {
  return -EINVAL;
}
```

i'd suggest rewrite this part usinf `std::filesystem()` for better readability.

File: src/blk/pmem/PMEMDevice.h
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.h:33]:
```suggestion
  bool devdax_device = false;
```

File: src/blk/pmem/PMEMDevice.cc
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:259]: wrong indent.

File: src/blk/pmem/PMEMDevice.cc
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:246]: please define variable close to where it is used for the first time for better readability.

File: src/blk/pmem/PMEMDevice.h
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.h:33]: ack

File: src/blk/pmem/PMEMDevice.cc
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:68]: ack

File: src/blk/pmem/PMEMDevice.cc
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:83]: Can we revise it later?

File: src/blk/pmem/PMEMDevice.cc
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:246]: ack

File: src/blk/pmem/PMEMDevice.cc
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:259]: ack

File: src/blk/pmem/PMEMDevice.cc
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:43]: now that this variable is initialized in-class, could just drop this line.

File: src/blk/pmem/PMEMDevice.cc
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:66]: please define a variable closer to where it is used for better readability.

File: src/blk/pmem/PMEMDevice.cc
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:98]:
```suggestion
  if (total_size != nullptr) {
```

more a là mode this way.

File: src/blk/pmem/PMEMDevice.cc
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:101]: this variable is assigned, but never read.

File: src/blk/pmem/PMEMDevice.cc
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:113]: please define a variable closer to where it is used for better readability.

File: src/blk/pmem/PMEMDevice.h
Comment (tchaikov) [reply src/blk/pmem/PMEMDevice.h:44]: might want to remove the trailing `;`.

File: src/blk/pmem/PMEMDevice.cc
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:43]: ack

File: src/blk/pmem/PMEMDevice.cc
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:66]: ack

File: src/blk/pmem/PMEMDevice.cc
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:98]: ack

File: src/blk/pmem/PMEMDevice.cc
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:113]: ack

File: src/blk/pmem/PMEMDevice.cc
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:101]: ack

File: src/blk/pmem/PMEMDevice.h
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.h:44]: ack

File: src/blk/pmem/PMEMDevice.cc
Comment (tchaikov) [reply src/blk/pmem/PMEMDevice.cc]: sure.

File: src/blk/pmem/PMEMDevice.cc
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:101]: wrong indent.

File: src/blk/pmem/PMEMDevice.cc
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:103]: wrong indent.

File: src/blk/pmem/PMEMDevice.cc
Comment (tchaikov) [inline src/blk/pmem/PMEMDevice.cc:147]: warp this line if you please.

**Review by tchaikov — COMMENTED** [review]
lgtm modulo some formatting nits.

File: src/blk/pmem/PMEMDevice.cc
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:91]: ack

File: src/blk/pmem/PMEMDevice.cc
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:94]: ack

File: src/blk/pmem/PMEMDevice.cc
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:101]: ack

File: src/blk/pmem/PMEMDevice.cc
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc:103]: ack

File: src/blk/pmem/PMEMDevice.cc
Comment (optimistyzy) [reply src/blk/pmem/PMEMDevice.cc]: ack

**Review by tchaikov — APPROVED** [review]

**Comment by tchaikov** [comment]
@majianpeng the change looks good to me. do you want to have a final review to it?

**Comment by tchaikov** [comment]
jenkins test api

**Comment by majianpeng** [comment]
LGTM.

**Comment by majianpeng** [comment]
jenkins test api

**Comment by optimistyzy** [comment]
jenkins test api

**Comment by tchaikov** [comment]
building at https://shaman.ceph.com/builds/ceph/wip-pr-45609-kefu/2b3a4d668874f2c6e4fff53777fc0d7c8af796eb/

**Comment by tchaikov** [comment]
being tested at https://pulpito.ceph.com/kchai-2022-04-19_13:32:12-rados-wip-pr-45609-kefu-distro-default-smithi/

**Comment by tchaikov** [comment]
failure tracked by https://tracker.ceph.com/issues/53342

File: src/blk/pmem/PMEMDevice.cc
Comment (tchaikov) [reply src/blk/pmem/PMEMDevice.cc]: #46122

**Pull request merged**
